File: src/main/java/com/github/oxyethylene/otelcollector/grpc/LogsServiceImpl.java
package com.github.oxyethylene.otelcollector.grpc;

import io.grpc.stub.StreamObserver;
import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceRequest;
import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceResponse;
import io.opentelemetry.proto.collector.logs.v1.LogsServiceGrpc;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class LogsServiceImpl extends LogsServiceGrpc.LogsServiceImplBase {
    private static final Logger log = LoggerFactory.getLogger(LogsServiceImpl.class);

    @Override
    public void export(ExportLogsServiceRequest request, StreamObserver<ExportLogsServiceResponse> responseObserver) {
        log.info("Received OTLP logs export");
        ExportLogsServiceResponse resp = ExportLogsServiceResponse.getDefaultInstance();
        responseObserver.onNext(resp);
        responseObserver.onCompleted();
    }
}

File: src/main/kotlin/com/github/oxyethylene/otelcollector/grpc/LogsServiceImpl.kt
package com.github.oxyethylene.otelcollector.grpc

import com.github.oxyethylene.otelcollector.model.StoredLog
import com.github.oxyethylene.otelcollector.storage.LogStore
import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceRequest
import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceResponse
import io.opentelemetry.proto.collector.logs.v1.LogsServiceGrpcKt
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class LogsServiceImpl(private val logStore: LogStore) : LogsServiceGrpcKt.LogsServiceCoroutineImplBase() {
    override suspend fun export(request: ExportLogsServiceRequest): ExportLogsServiceResponse {
        request.resourceLogsList.forEach { resourceLogs ->
            val resource = resourceLogs.resource
            resourceLogs.scopeLogsList.forEach { scopeLogs ->
                val scopeName = scopeLogs.scope.name
                scopeLogs.logRecordsList.forEach { logRecord ->
                    logStore.store(StoredLog.fromProto(logRecord, resource, scopeName))
                }
            }
        }
        log.debug("Received OTLP logs export")
        return ExportLogsServiceResponse.getDefaultInstance()
    }

    companion object {
        private val log = LoggerFactory.getLogger(LogsServiceImpl::class.java)
    }
}
